<launch>

  <!-- Name of the camera we will be streaming data from -->
  <arg name="camera" />
  <arg name="rate" default="60" />
  <!-- Color that will be proccesed. TODO: Use all colors -->
  <arg name="color" default="blaze_orange" />
  <arg name="immediate_tracking" default="true" />

  <!-- Stage 1: Downscale-->
  <!-- <include file="$(find image_transforms)/launch/image_scaler.launch" > -->
  <!--   <arg name="image_in" value="$(arg camera)/image_color" /> -->
  <!--   <arg name="image_out" value="$(arg camera)/image_color_scaled" /> -->
  <!--   <arg name="info_in" value="$(arg camera)/camera_info" /> -->
  <!--   <arg name="info_out" value="$(arg camera)/camera_info_scaled" /> -->
  <!--   <arg name="scale" value="0.5" /> -->
  <!-- </include> -->

  <!-- Stage 2: Color Classifier -->
  <remap from="color_classifier/image_color" to="$(arg camera)/image_color_scaled" />
  
  <include file="$(find color_classification)/launch/color_classifier.launch" >
    <arg name="rate" value="$(arg rate)" />
  </include>

  <!-- Stage 3: Shape Matcher -->
  <remap from="shape_matcher/image_mono" to="color_classifier/$(arg color)_classified" />

  <include file="$(find shape_matching)/launch/shape_matcher.launch" >
    <arg name="rate" value="$(arg rate)" />
  </include>

  <!-- Stage 4: Object Tracker -->
  <remap from="unimodal_object_tracker/matched_shapes" to="shape_matcher/matched_shapes" />
  <remap from="unimodal_object_tracker/camera_info" to="$(arg camera)/camera_info_scaled" />
  
  <include file="$(find object_tracking)/launch/unimodal_object_tracker.launch" >
    <arg name="rate" value="$(arg rate)" />
    <arg name="immediate_tracking" value="$(arg immediate_tracking)" />
  </include>

  <!-- Params -->
  <include ns="model" file="$(find color_model)/launch/upload_colors.launch" />
  <include ns="model" file="$(find shape_model)/launch/upload_shapes.launch" />  
  <include ns="model" file="$(find object_model)/launch/upload_objects.launch" />
  
  <!-- TODO: Don't load this here -->
  <rosparam command="load" file="$(find shape_matching)/params/image_proc.yaml" ns="shape_matcher/image_proc" />

</launch>